File: lc/linkedlist/java/ReverseList.java
package lc.linkedlist;

public class ReverseList {
    public ListNode reverse(ListNode head) {
        
        if(head == null || head.next == null) return head;
        
        ListNode prev = null;
        
        while(head != null) {
        	
            ListNode tmp = head.next;
            head.next = prev;
            prev = head;
            head = tmp;
            
        }
        
        return prev;
    }
}

File: lc/linkedlist/ruby/reverse_list.rb
def reverse_list(head)
    dummy_head = ListNode.new 0
    dummy_head.next = head

    # if ruby >= 2.3, just use safe operator like 'head&.next'
    while head && head.next
        tmp = head.next
        head.next = tmp.next
        tmp.next = dummy_head.next
        dummy_head.next = tmp
    end

    return dummy_head.next
end